File: python/DataStructures/Stack.py
"""
Stack.py

Stack (last-in-first-out collection) - linked-list implementation 

Implements:
.push() 	| O(1)
.pop()		| O(1)
.is_empty()	| O(1)

"""
class Node:
	def __init__(self, data, next=None):
		self.data  = data 
		self.next = next

class Stack:

	def __init__(self):
		self.head = None

	def push(self, data):
		if self.is_empty():
			self.head = Node(data)
		else:
			oldhead = self.head
			self.head = Node(data, oldhead)

	def pop(self):
		oldhead = self.head
		self.head = self.head.next
		return oldhead

	def is_empty(self):
		return self.head == None

	def __str__(self):
		"""
		Prints the current list in the form of a Python list            
		"""
		current = self.head
		toPrint = []
		while current != None:
			toPrint.append(current.data)
			current = current.next
		return str(toPrint) 


if __name__ == "__main__":
	s = Stack()
	for i in range(0,20):
		s.push(i)
	s.pop()
	s.pop()
	print(s)

File: java/src/DataStructures/Stack.java
package DataStructures;
/**
 * Stack - LIFO queue
 * 		 - Linked List implementation
 * 		 - constant time push() and pop() methods
 * @author mashabelyi
 *
 * @param <Type>
 */

public class Stack<Type> {
	
	private Node head;
	private int size;
	
	/**
	 * Linked List Node implementation
	 *
	 * @param <Type>
	 */
	private static class Node<Type> {
		private Type data;
		private Node<Type> next;
		public Node(Type data, Node next) {
			this.data = data;
			this.next = next;
		}
	}
	
	public Stack() {
		this.size = 0;
		this.head = null;
	}
	
	/**
	 * push()
	 * 
	 * Insert at beginning of list 
	 * O(1)
	 * @param data - value to insert
	 */
	public void push(Type data) {
		if (isEmpty()) {
			this.head = new Node(data, null);
		}
		else {
			Node oldhead = this.head;
			this.head = new Node(data, oldhead);
		}
		size++;
	}
	
	/**
	 * pop()
	 * 
	 * Remove from beginning of list
	 * O(1)
	 * @return - removed Node
	 */
	public Node pop() {
		if (isEmpty()) {
			return null;
		}
		else {
			Node oldhead = this.head;
			this.head = this.head.next;
			size--;
			return oldhead;
		}
		
	}
	
	/**
	 * isEmpty()
	 * 
	 * @return
	 */
	public Boolean isEmpty() {
		return this.head == null;
	}
	
	public String toString() {
		String[] toPrint = new String[size];
		Node tmp = this.head;
		int i = 0;
		while (tmp != null) {
			toPrint[i] = (String) tmp.data;
			tmp = tmp.next;
			i++;
		}
		return String.join(" ", toPrint);
	}
	
	public static void main(String[] args) {
		Stack<String> s = new Stack<String>();
		String input = "a b c d e f g h i j k l m n o p";
		for (String str: input.split(" ")) {
			s.push(str);
		}
		System.out.println(s);
		s.pop();
		s.pop();
		System.out.println(s);
		
	}
}
